scipy/spatial/src/distance_impl.h

```diff
@@ -39,11 +39,12 @@ _row_norms(const double *X, npy_intp num_rows, const npy_intp num_cols, double *
     /* Compute the row norms. */
     npy_intp i, j;
     for (i = 0; i < num_rows; ++i) {
-        for (j = 0; j < num_cols; ++j, ++X) {
-            const double curr_val = *X;
-            norms_buff[i] += curr_val * curr_val;
+        double accumulator = 0.0;
+        for (j = 0; j < num_cols; ++j) {
+            const double curr_val = X[(i * num_cols) + j];
```

Comment:
I'm with Ilan on the topic of pointer arithmetic vs. indexing an array. In fact, the book I learned C from, Stephen Kochan's "Programming in C" 4th Edition (2014),  states in "Chapter 10: Pointers":

> ... the process of indexing an array takes more time to execute than does the process of accessing the contents of a pointer. In fact, this is one of the main reasons why pointers are used to access the elements of an array--the code that is generated is generally more efficient.

Maybe the gap has closed more with modern compilers, but still.

Also, this branch doesn't cross the default significance threshold for our `asv` benchmarks to report a performance improvement:

`asv continuous -e -b ".*Xdist.*" main DavidM603/main`

`BENCHMARKS NOT SIGNIFICANTLY CHANGED.`

So, on my side, I'm not really hugely in favor of mutating the code (which always carries *some* risk) until we cross that threshold or add a benchmark that would. I still remember reading that book 10 years ago when it was raining constantly in Oxford...

Reply:
It's pretty fair to err on the side of "if it ain't broke, don't fix it", especially when someone new shows up and pokes at old code to marginal effect. I'm suggesting changes based on disassembly and benchmarks, not guesswork, but I should really put those things together in a presentable way and share them instead of doing it all ad-hoc. Very new to open source, sorry about that, I'll present some codegen change examples and microbench numbers over the weekend.

One question before working on that though: This PR makes it so that _row_norms no longer reads from the output buffer; it doesn't need to be zeroed. Would you be categorically opposed to switching to malloc for these buffers? If not, I want to include that in the PR and run the circleci benchmarks again.

Reply:
If the complexity and portability remain the same and we can actually see the performance change in our standard `asv` benchmark suite (without needing to lower the threshold of change), it may be worth considering.

If it requires microbenchmarks and dissecting specific cases on specific architectures outside of our regular `asv` benchmarks, I'm less enthusiastic about the tradeoffs on review time, etc.

Reply:
At 1-3 % on some architectures, I'd say it is not worth the churn, even if the diff remains as small as it is now.

Reply:
In that case I'm just going to drop it and work on other things.